ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/HardFork/OracularClock.hs

```diff
@@ -0,0 +1,242 @@
+-- correspondence in the absence of a Common Prefix violation. So this clock
+-- can indeed know /the/ future.
+data OracularClock m = OracularClock
+    { -- | Returns 'False' if the requested slot is already over
+      --
+      -- Note: blocks indefinitely if clock becomes exhausted before requested
+      -- slot.
+      blockUntilSlot :: SlotNo -> m Bool
+
+      -- | The current delay duration until the onset of the next slot
+      --
+      -- Note: blocks indefinitely if the clock is already exhausted.
+    , delayUntilNextSlot :: m NominalDiffTime
+
+      -- | A mock system time
+      --
+      -- Note: this 'BTime.systemTimeCurrent' blocks indefinitely if clock is
+      -- already exhausted.
+    , finiteSystemTime :: BTime.SystemTime m
+
+      -- | Convert 'Control.Monad.Class.MonadTime.getCurrentTime' to 'SlotNo'
+      --
+      -- INVARIANT: In @io-sim@, the @getCurrentTime@ use (by default) the
+      -- monotonic clock, and it advances /before/ threads correspondingly
+      -- wake-up from a @threadDelay@.
+      --
+      -- So we rely on the (monotonic) clock instead of using a ticker thread
+      -- and internal state, since a calling thread and the ticker thread might
+      -- wake up simultaneously from a
+      -- 'Control.Monad.Class.MonadTimer.threadDelay' call: it's ambiguous
+      -- which thread would win the race to respectively read or increment the
+      -- internal state.
+      --
+      -- Note: we do use a ticker thread for 'onSlotChange'.
+      --
+      -- Note: blocks indefinitely if the clock is already exhausted.
+    , getCurrentSlot :: m SlotNo
+
+      -- | See 'onSlotChange'
+    , onSlotChange_ :: HasCallStack
+                    => ResourceRegistry m
+                    -> String
+                    -> (SlotNo -> m ())
+                    -> m (m ())
+
+      -- | Block until the clock is exhausted
+    , waitUntilDone :: m ()
+    }
+
+-- | Forks a thread that executes an action at the onset of each slot
+--
+-- Returns an action that cancels the thread.
+--
+-- INVARIANT: In @io-sim@, there is no race:
+-- @'finiteSystemTime'.systemTimeCurrent@ and hence 'getCurrentSlot' called
+-- from within the given action will always return the correct slot.
+--
+-- See the discussion of ticker threads in 'getCurrentSlot'.
+onSlotChange :: HasCallStack
+             => ResourceRegistry m
+             -> OracularClock m
+             -> String
+             -> (SlotNo -> m ())
+             -> m (m ())
+onSlotChange reg clk = onSlotChange_ clk reg
+    -- jumping the hoop so HasCallStack is useful
+
+-- | Perform a " transaction " within the current slot
+--
+-- The callback should be idempotent.
+--
+-- We don't export a 'getCurrentSlotSTM' function because that leads to races
+-- with 'threadDelay'. This function repeatedly attempts to execute the
+-- callback until it is able to do so with 'getCurrentSlot' returning the same
+-- provided value before and after.
```

Comment:
Is this comment up to date? There is nothing executed repeatedly here is there?

Reply:
Ah, yes, they're stale. I'll refine them.

Reply:
I added one more variable to `Network.hs` for explicit synchronization, so that this function is not unnecessary. I removed it.